src/szk.rs

```diff
@@ -0,0 +1,565 @@
+            proof_share_seed_and_blind: helper_seed.clone(),
+            joint_rand_part: helper_joint_rand_part,
+        };
+        Ok([leader_proof_share, helper_proof_share])
+    }
+
+    fn query(
+        &self,
+        input_share: &[T::Field],
+        proof_share: SzkProofShare<T::Field, SEED_SIZE>,
+        verify_key: &[u8; SEED_SIZE],
+        nonce: &[u8; 16],
+    ) -> Result<SzkQueryShare<T::Field, SEED_SIZE>, SzkError> {
+        let query_rand = self.derive_query_rand(verify_key, nonce);
+
+        let is_leader = match proof_share {
+            SzkProofShare::Leader { .. } => true,
+            SzkProofShare::Helper { .. } => false,
+        };
+
+        let flp_proof_share = match proof_share {
+            SzkProofShare::Leader {
+                ref uncompressed_proof_share,
+                ..
+            } => Cow::Borrowed(uncompressed_proof_share),
+            SzkProofShare::Helper {
+                ref proof_share_seed_and_blind,
+                ..
+            } => Cow::Owned(self.derive_helper_proof_share(proof_share_seed_and_blind)),
+        };
+
+        let (joint_rand, jr_seed_and_part) = if self.typ.joint_rand_len() > 0 {
+            let (seed, peer_joint_rand_part) = match proof_share {
+                SzkProofShare::Leader {
+                    uncompressed_proof_share: _,
+                    blind_and_joint_rand_part,
+                } => blind_and_joint_rand_part.expect("jr_len > 0 so there is joint_randomness"),
+                SzkProofShare::Helper {
+                    proof_share_seed_and_blind,
+                    joint_rand_part,
+                } => (
+                    proof_share_seed_and_blind,
+                    joint_rand_part.expect("jr_len > 0 so there is joint randomness"),
+                ),
+            };
+            let host_joint_rand_part = self.derive_joint_rand_part(&seed, input_share, nonce)?;
+            let (leader_joint_rand_part, helper_joint_rand_part) = if is_leader {
+                (&host_joint_rand_part, &peer_joint_rand_part)
+            } else {
+                (&peer_joint_rand_part, &host_joint_rand_part)
+            };
+            let (joint_rand_seed, joint_rand) =
+                self.derive_joint_rand(leader_joint_rand_part, helper_joint_rand_part);
+            (joint_rand, Some([joint_rand_seed, host_joint_rand_part]))
+        } else {
+            (Vec::new(), None)
+        };
+        let verifier_share = self.typ.query(
+            input_share,
+            flp_proof_share.as_ref(),
+            &query_rand,
+            &joint_rand,
+            2,
+        )?;
+        Ok(SzkQueryShare {
+            joint_rand_seed_and_part: jr_seed_and_part,
+            verifier: verifier_share,
+        })
+    }
+
+    /// Returns true if the verifier message indicates that the input from which
+    /// it was generated is valid.
+    fn decide(
+        &self,
+        verifier: &[T::Field],
+        l_jr_part: Option<Seed<SEED_SIZE>>,
```

Comment:
nit: In general, it's a good idea to use consistent variable names across function calls. This make it easier for the reader to identify related concepts across the code.

Here and below.
```suggestion
        leader_joint_rand_part: Option<Seed<SEED_SIZE>>,
```